README.md
LEB128
==========

A library helping to deal with Little Endian Base 128.

master:
[![Build Status](https://travis-ci.org/yageek/LEB128.svg?branch=master)](https://travis-ci.org/yageek/LEB128)
develop:
[![Build Status](https://travis-ci.org/yageek/LEB128.svg?branch=develop)](https://travis-ci.org/yageek/LEB128)

[![MIT License](http://img.shields.io/badge/license-MIT-blue.svg?style=flat)](LICENSE)
[![Carthage compatible](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat)](https://github.com/Carthage/Carthage)
[![CocoaPods Compatible](https://img.shields.io/cocoapods/v/LEB128.svg)](https://img.shields.io/cocoapods/v/LEB128.svg)
[![Platform](https://img.shields.io/cocoapods/p/LEB128.svg?style=flat)](http://cocoadocs.org/docsets/LEB128)

# Install

## Carthage

```
github "yageek/LEB128"
```

## Swift Package Manager

```
import PackageDescription

let package = Package(
    name: "myPackage",
    dependencies: [
    .Package(url: "https://github.com/yageek/LEB128.git", majorVersion: 1, minor: 0)
])
```

## Cocoapods

```
pod 'Leb128'
```

# Usage

```swift

    /// Encoding
    let buff = ByteBuffer(size: 5)
    let length = encodeUnsignedLEB(buff, value: value)
    print("Value: \(buff[0..<length])")
    

    let value: Int = 16256
    let buff = ByteBuffer(size: 5)
    let length = encodeSignedLEB(buff, value: value)
    print("Value: \(buff[0..<length])")


    /// Decoding

    let encodedSigned = decodeUnsignedLEB(ByteBuffer(elements:[0x80, 0x7f]))
    print("Value: \(encodedSigned)")

    let encodedUSigned = decodeSignedLEB(ByteBuffer(elements:[0x80, 0x7f]))
    print("Value: \(encodedUSigned)")

```

# License

LEB128 is released under the MIT license. See LICENSE for details.


CHANGELOG.md
# Change Log

## [2.0.0](https://github.com/yageek/LEB128/tree/2.0.0) (2016-11-01)
[Full Changelog](https://github.com/yageek/LEB128/compare/1.0.0...2.0.0)
- Add compatibility with Swift 3.0

**Fixed bugs:**

- Decode Signed Integer [\#1](https://github.com/yageek/LEB128/issues/1)



\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

Tests/LEB128Tests/LEB128Tests.swift
//
//  LEB128Tests.swift
//  LEB128Tests
//
//  Created by Yannick Heinrich on 16.06.16.
//
//

import XCTest
import LEB128


class LEB128Tests: XCTestCase {

    let unsignedMap: [UInt: Array<Byte>] = [
        0: [0],
        1: [0x01],
        2: [0x02],
        127: [0x7f],
        128: [0x80, 0x01],
        129: [0x81, 0x01],
        130: [0x82, 0x01],
        12857: [0xB9, 0x64],
        16256: [0x80, 0x7f],
        0x40c: [0x8c, 0x08],
        624485: [0xe5, 0x8e, 0x26],
    ]

    let signedMap: [Int: Array<Byte>] = [
        0: [0],
        1: [0x01],
        2: [0x02],
        127:  [0xFF, 0x00],
        128:  [0x80, 0x01],
        129:  [0x81, 0x01],
        -1: [0x7f],
        -2:   [0x7e],
        -127: [0x81, 0x7f],
        -128: [0x80, 0x7f],
        -129: [0xFF, 0x7e],
        ]


    // MARK: Helpers
    private func encodeUnsignedLeb(leb: UInt) -> ArraySlice<Byte> {
        let buff = ByteBuffer(size: 5)
        let length = encodeUnsignedLEB(buff, value: leb)
        return buff[0..<length]
    }

    private func encodeSignedLeb(leb: Int) -> ArraySlice<Byte> {
        let buff = ByteBuffer(size: 5)
        let length = encodeSignedLEB(buff, value: leb)
        return buff[0..<length]
    }


    // MARK: Unsigned integer

    func testUnsignedEncode() {
        for (raw, bytes) in unsignedMap {
            let expected = ByteBuffer(elements: bytes)
            XCTAssertEqual(expected[0..<expected.size], encodeUnsignedLeb(leb: raw))
        }
    }

    func testDecodeUnsignedLeb() {

        for (raw, bytes) in unsignedMap {
            let elements = ByteBuffer(elements: bytes)
            XCTAssertEqual(raw, decodeUnsignedLEB(elements))
        }
    }


    // MARK: Signed integer

    func testSignedEncode() {
        for (raw, bytes) in signedMap {
            let expected = ByteBuffer(elements: bytes)
            XCTAssertEqual(expected[0..<expected.size], encodeSignedLeb(leb: raw))
        }

    }

    func testDecodSignedLeb() {

        for (raw, bytes) in signedMap {
            let elements = ByteBuffer(elements: bytes)
            XCTAssertEqual(raw, decodeSignedLEB(elements))
        }
    }


}


Sources/LEB128.swift
/**
    Algorithms are taken from
   "DWARF Debugging Information Format Specification Version 2.0, Draft" (PDF)
 */

// MARK: - Core

/// Typealias to byte
public typealias Byte = UInt8

/// Protocol for output buffer
public protocol ByteOut {
    func write(_ byte: UInt8)
}

/// Protocol for input buffer
public protocol ByteIn {
    func read() -> UInt8
}

// MARK: - Unsigned Integer
/**
 Encode the unsigned integer as LEB128

 - parameter out: The `ByteOut` to write in
 - parameter value: The `UInt` value to convert
 - returns:         The number of written byte
 */

public func encodeUnsignedLEB(_ out: ByteOut, value: UInt) -> Int {
    var value = value
    var count = 0

    repeat {
        var byte = Byte(value & 0x7F)
        value = value >> 7
        if value != 0 {
            byte |= 0x80
        }
        out.write(byte)
        count += 1
    } while value != 0

    return count
}

/**
 Decode the unsigned integer as LEB128

 - parameter input: The `ByteIn` to read
 - returns:         The decoded value
 */
public func decodeUnsignedLEB(_ input: ByteIn) -> UInt {
    var result: UInt = 0
    var shift: UInt = 0

    while true {
        let byte = input.read()
        result |= ((UInt(byte) & 0x7F) << shift)

        if (byte >> 7) == 0 {
            break
        }
        shift += 7
    }
    return result
}

// MARK: - Signed Integer

/**
 Encode the signed integer as LEB128

 - parameter out: The `ByteOut` to write in
 - parameter val: The `Int` value to convert
 - returns:         The number of written byte
 */

public func encodeSignedLEB(_ out: ByteOut, value: Int) -> Int {

    var value = value
    var more = true
    var count = 0

    while more {
        var byte = Byte(value & 0x7F)
        value = value >> 7

        if (value == 0 && (byte >> 6) == 0) || (value == -1 && (byte >> 6) == 1) {
            more = false
        } else {
            byte |= 0x80
        }

        out.write(byte)
        count += 1
    }
    return count
}


/**
 Decode the unsigned integer as LEB128

 - parameter input: The `ByteIn` to read
 - returns:         The decoded value
 */
public func decodeSignedLEB(_ input: ByteIn) -> Int {
    var result: Int = 0
    var shift: Int = 0
    let size: Int = MemoryLayout<Int>.size * 8
    var byte: Byte = 0

    while true {
        byte = input.read()
        result |= ((Int(byte) & 0x7F) << shift)
        shift += 7

        if ((byte & 0x80) >> 7) == 0 {
            break
        }

    }

    if (shift < size) && ((Int(byte) & 0x40) >> 6) == 1 {
        result |= -(1 << shift)
    }
    return result
}


Sources/ByteBuffer.swift
//
//  ByteBuffer.swift
//  LEB128
//
//  Created by Yannick Heinrich on 16.06.16.
//
//

import Foundation

public func == (left: ByteBuffer, right: ByteBuffer) -> Bool {
    guard left.size == right.size else { return false }

    for i in 0..<left.size {
        if left[i] != right[i] {
            return false
        }
    }
    return true
}

/// ByteBuffer holds a buffer.
open class ByteBuffer: ByteOut, ByteIn, Equatable {

    fileprivate var buff: ContiguousArray<Byte>
    fileprivate var index: Int

    /**
      Returns the buffer size
      - returns: The size of the buffer
     */
    open var size: Int { return buff.count }

    /**
      Initialize a buffer with size

     - parameter size: Size as Int
    */
    public init(size: Int) {
        index = 0
        buff = ContiguousArray(repeating: 0, count: size)
    }

    /**
     Convenience init for a readable buffer.

     - parameter elements: Elements as `Array<Byte>`
     */
    public convenience init(elements: Array<Byte>) {
        self.init(size: elements.count)

        for element in elements {
            write(element)
        }

        index = 0
    }


    /**
        Reset the buffer
    */
    open func clear() {
        buff = ContiguousArray<Byte>(repeating: 0, count: size)
        index = 0
    }

    /**
      Write a `Byte` in the buffer

      - parameter byte: The value to write in
     */
    open func write(_ byte: Byte) {
        defer {
            index += 1
        }
        buff[index] = byte

    }

    /**
      Read the current value

     - returns: The `Byte` at the current index.
   */
    open func read() -> Byte {
        defer {
            index += 1
        }
        return buff[index]
    }

    open subscript (index: Int) -> Byte {
        get {
            return buff[index]
        }
    }

    open subscript(range: Range<Int>) -> ArraySlice<Byte> {
        get {
            return buff[range]
        }
    }
}


LEB128Tests/LEB128Tests.swift
//
//  LEB128Tests.swift
//  LEB128Tests
//
//  Created by Yannick Heinrich on 16.06.16.
//
//

import XCTest
import LEB128


class LEB128Tests: XCTestCase {

    let unsignedMap: [UInt: Array<Byte>] = [
        0: [0],
        1: [0x01],
        2: [0x02],
        127: [0x7f],
        128: [0x80, 0x01],
        129: [0x81, 0x01],
        130: [0x82, 0x01],
        12857: [0xB9, 0x64],
        16256: [0x80, 0x7f],
        0x40c: [0x8c, 0x08],
        624485: [0xe5, 0x8e, 0x26],
    ]

    let signedMap: [Int: Array<Byte>] = [
        0: [0],
        1: [0x01],
        2: [0x02],
        127:  [0xFF, 0x00],
        128:  [0x80, 0x01],
        129:  [0x81, 0x01],
        -1: [0x7f],
        -2:   [0x7e],
        -127: [0x81, 0x7f],
        -128: [0x80, 0x7f],
        -129: [0xFF, 0x7e],
        ]


    // MARK: Helpers
    private func encodeUnsignedLeb(leb: UInt) -> ArraySlice<Byte> {
        let buff = ByteBuffer(size: 5)
        let length = encodeUnsignedLEB(buff, value: leb)
        return buff[0..<length]
    }

    private func encodeSignedLeb(leb: Int) -> ArraySlice<Byte> {
        let buff = ByteBuffer(size: 5)
        let length = encodeSignedLEB(buff, value: leb)
        return buff[0..<length]
    }


    // MARK: Unsigned integer

    func testUnsignedEncode() {
        for (raw, bytes) in unsignedMap {
            let expected = ByteBuffer(elements: bytes)
            XCTAssertEqual(expected[0..<expected.size], encodeUnsignedLeb(raw))
        }
    }

    func testDecodeUnsignedLeb() {

        for (raw, bytes) in unsignedMap {
            let elements = ByteBuffer(elements: bytes)
            XCTAssertEqual(raw, decodeUnsignedLEB(elements))
        }
    }


    // MARK: Signed integer

    func testSignedEncode() {
        for (raw, bytes) in signedMap {
            let expected = ByteBuffer(elements: bytes)
            XCTAssertEqual(expected[0..<expected.size], encodeSignedLeb(raw))
        }

    }

    func testDecodSignedLeb() {

        for (raw, bytes) in signedMap {
            let elements = ByteBuffer(elements: bytes)
            XCTAssertEqual(raw, decodeSignedLEB(elements))
        }
    }


}
